File: 560-subarray-sum-equals-k/subarray-sum-equals-k.py
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        n = len(nums)
        d = {0: 1}
        curr_sum = 0
        count = 0

        for num in nums:
            curr_sum += num

            if curr_sum - k in d:
                count += d[curr_sum - k]
            
            d[curr_sum] = d.get(curr_sum, 0) + 1


        return count

File: 560-subarray-sum-equals-k/subarray-sum-equals-k.cpp
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        int currentSum = 0, prefixSum = 0, res = 0;
        unordered_map<int, int> mpp;

        mpp[0] = 1;

        for(auto num : nums) {
            currentSum += num;
            prefixSum = currentSum - k;
            res += mpp[prefixSum];
            mpp[currentSum] += 1;
        }
        return res;
    }
};